doc/mgr/modules.rst

```diff
@@ -197,6 +205,192 @@ when they are sent:
+           return {"gravitons": 999, "tachyons": 404}
+
+   @CLICommand('antigravity list wormholes', perm='r')
+   @Responder()
+   def list_wormholes(self, oid: str, details: bool = False) -> MyCleverObject:
+       '''List wormholes associated with the supplied oid.
+       '''
+       ...
+
+The behavior of the automatic output formatting can be customized and extednted
+to other types of formatting (XML, Plain Text, etc). As this is a complex
+topic, please refer to the module documentation for the ``object_format``
+module.
+
+
+
+Error Handling
+++++++++++++++
+
+Additionally, the ``Responder`` decorator can automatically handle converting
+some  exceptions into response tuples. Any raised exception inheriting from
+``ErrorResponseBase`` will be automatically converted into a response tuple.
+The common approach will be to use ``ErrorResponse``, an exception type that
+can be used directly and has arguments for the error output and return value or
+it can be constructed from an existing exception using the ``wrap``
+classmethod. The wrap classmethod will automatically use the exception text and
+if available the ``errno`` property of other exceptions.
+
+Converting our previous example to use this exception handling approach:
+
+.. code:: python
+
+   @CLICommand('antigravity list wormholes', perm='r')
+   @Responder()
+   def list_wormholes(self, oid: str, details: bool = False) -> List[Dict[str, Any]]:
+       '''List wormholes associated with the supplied oid.
+       '''
+       try:
+           with self.open_wormhole_db() as db:
+               wormholes = db.query(oid=oid)
+       except UnknownOIDError:
+            raise ErrorResponse(f"Unknown oid: {oid}", return_value=-errno.ENOENT)
+       except WormholeDBError as err:
+           raise ErrorResponse.wrap(err)
+       if not details:
+           return [{'name': wh.name} for wh in wormholes]
+       return [{'name': wh.name, 'age': wh.get_age(), 'destination': wh.dest}
+               for wh in wormholes]
+
+
+.. note:: Because the decorator can not determine the difference between a
+   programming mistake and an expected error condition it does not try to
+   catch all exceptions.
+
+
+
+Additional Decorators
++++++++++++++++++++++
+
+The ``object_format`` module provides additional decorators to complement
+``Responder`` but for cases where ``Responder`` is insufficient or too "heavy
+weight".
+
+The ``ErrorResponseHandler`` decorator exists for cases where you *must* still
+return a manager response tuple but want to handle errors as exceptions (as in
+typical Python code). In short, it works like ``Responder`` but only with
+regards to exceptions. Just like ``Responder`` it handles exceptions that
+inherit from ``ErrorResponseBase``. This can be useful in cases where you need
+to return raw data in the output. Example:
+
+.. code:: python
+
+   @CLICommand('antigravity dump config', perm='r')
+   @ErrorResponseHandler()
```

Comment:
> Instead of creating a new decorator (that kinda overlaps with the `Responder` functionality), why not extending `Responder(...)` with a `return_tuple=True` argument or something?  The more helpers devels have to import and deal with, the less likely they'll use them (or the more likely they'll misuse them) disappointed

I'm not sure I agree with that, I agree that cognitive burden can be a thing, but given that it's a matter of selecting from three decorators (not 10s or 100s of choices) and they're all in the same module. I also feel that there's a (Single Responsibility Principle)[https://en.wikipedia.org/wiki/Single-responsibility_principle] argument to be made towards having multiple decorators.

Now, if the names are really lousy and hard to remember that's a different story - so I'm open to name change suggestions if it seems like the names don't correlate to what the decorators do.

> 
> And one step forward, does a devel have to explicitly enable any decorator or argument? The `Responder()` decorator could easily check if the return value is of `HandleCommandResult` type (we may stick to this convention: while there's lots of legacy code still using plain tuples for it, that was added for enforcing proper return format).
> 
> I remember we already talked about embedding the `Responder` functionality inside the `CLICommand` helper. While I'm not always fond of "automagic" things, once you already have one (like the `CLICommand`) I think it's preferable to extend (and keep hidden) it's magic than to add new helpers that people need to remind invoking. Only thing is to ensure that new magic doesn't break anything, but that can be ensured with some conservative approach (e.g.: checking return types for `HandleCommandResult` or for objects implementing the `SimpleDataProvider` protocol).

I'd prefer to wait and see how things go with this simpler and explicit form first. Once we have more experience we could make a "super-decorator" that wraps up more of the features together. 

Another argument for at least keeping the lower-level decorators is that while I was fixing the docs I learned about the older `COMMANDS` attribute. Looking at that code that used that approach the decorators could be used on functions in that older way too if needed (modulo needing an explicit format flag - a topic for another doc update ;-) ) .  So there's a composability argument to be had for keeping them too, IMO.

Reply:
ping @epuertat

Reply:
@phlogistonjohn @adk3798 no objections from my side. I was mostly concerned about how to increase awareness/adoption of this.